tests/unit/test_library_saml.py

```diff
@@ -108,39 +108,7 @@ def test_saml_relation_data_to_relation_data():
     assert relation_data == expected_relation_data
 
 
-def test_requirer_charm_does_not_emit_event_id_no_leader():
-    """
-    arrange: set up a charm with no leadership.
-    act: trigger a relation changed event.
-    assert: no events are emitted.
-    """
-    relation_data = {
-        "entity_id": "https://login.staging.ubuntu.com",
-        "metadata_url": "https://login.staging.ubuntu.com/saml/metadata",
-        "x509certs": "cert1,cert2",
-        "single_sign_on_service_redirect_url": "https://login.staging.ubuntu.com/saml/",
-        "single_sign_on_service_redirect_binding": (
-            "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect"
-        ),
-        "single_logout_service_redirect_url": "https://login.staging.ubuntu.com/+logout",
-        "single_logout_service_redirect_binding": (
-            "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect"
-        ),
-    }
-    harness = Harness(SamlRequirerCharm, meta=REQUIRER_METADATA)
-    harness.begin()
-    harness.set_leader(False)
-    relation_id = harness.add_relation("saml", "saml-provider")
-    harness.add_relation_unit(relation_id, "saml-provider/0")
-    harness.update_relation_data(
-        relation_id,
-        "saml-provider",
-        relation_data,
-    )
-    assert len(harness.charm.events) == 0
```

Comment:
We're removing a test that it's only emitted on the leader but we're not adding a test that it is emitted on non-leaders I think?

Reply:
We already have a test to check if the event is emitted

Reply:
You have one that fires on the leader, but not on a non-leader if I'm not mistaken?

Reply:
It doesn't mattter, it's the same flow

Reply:
It does matter, if we add back in `if not self.charm.unit.is_leader()` this would still pass. The fact that you've had to change the functionality to me means it's something worth testing.

Reply:
Yeh definitely, especially since we seem to need to change this behaviour

Reply:
Well, I agree with both of you - if I understand correctly, now we emit events regardless of whether the unit is a leader or not - in that way, we don't need the test as is, it should be removed. Now to guard ourselves against a regression if/when we add a condition for leadership, there should be a test, right? If we test that regardless of a leadership an event is emitted, that could go in one test to avoid duplication - test once when we are the leader and once when we are not. That will be a pretty beafy unit test, though if we split it, it will be mostly the same code. Perhaps parameterize this so that with the same data and different leadership we get the same result with the current code? Removing code is always nice and I think making sure a regression doesn't happen is good to have as well.

Reply:
I agree we need to test both case, and renaming the other test is not right as in this other test we explicitely set the leadership to "True"
If we had an issue explicitely due to leadership, we need to have tests that covers with leadership and without, otherwise as already mentioned we may have a regression on either leader or non leader.

A parametrized test is probably the best as we have the same test for both cases